src/crimson/os/seastore/onode_manager/simple-fltree/onode_node.h

```diff
@@ -0,0 +1,962 @@
+  // XXX: magic numbers here
+  static constexpr bool item_in_key = (N == 3 &&
+                                       NodeType == ntype_t::inner);
+  using partial_key_t = std::conditional_t<N < 3,
+                                           variable_sized_key_t,
+                                           empty_key_t>;
+
+  std::pair<const key_t&, const partial_key_t&>
+  key_at(unsigned slot) const;
+
+  // update an existing oid with the specified item
+  ghobject_t update_oid_with_slot(unsigned slot, const ghobject_t& oid) const;
+  const_item_t item_at(const key_t& key) const;
+  void dump(std::ostream& os) const;
+
+  // for debugging only.
+  static constexpr bool is_leaf() {
+    return node_type == ntype_t::leaf;
+  }
+
+  bool _is_leaf() const {
+    return tag.is_leaf();
+  }
+
+  char* from_end(uint16_t offset);
+  const char* from_end(uint16_t offset) const;
+  uint16_t used_space() const;
+  uint16_t free_space() const {
+    return capacity() - used_space();
+  }
+  static uint16_t capacity();
+  // TODO: if it's allowed to update 2 siblings at the same time, we can have
+  //       B* tree
+  static constexpr uint16_t min_size();
+
+  // @return <minimum bytes to grab, maximum bytes to grab>
+  static constexpr std::pair<int16_t, int16_t> bytes_to_grab(uint16_t size);
+
+  // @return <minimum bytes to push, maximum bytes to push>
+  static constexpr std::pair<int16_t, int16_t> bytes_to_push(uint16_t size);
+
+  size_state_t size_state(uint16_t size) const;
+  bool is_underflow(uint16_t size) const;
+  int16_t size_with_key(unsigned slot, const ghobject_t& oid) const;
+  ordering_t compare_with_slot(unsigned slot, const ghobject_t& oid) const;
+  /// return the slot number of the first slot that is greater or equal to
+  /// key
+  std::pair<unsigned, bool> lower_bound(const ghobject_t& oid) const;
+  static uint16_t size_of_item(const ghobject_t& oid, const item_t& item);
+  bool is_overflow(const ghobject_t& oid, const item_t& item) const;
+  bool is_overflow(const ghobject_t& oid, const OnodeRef& item) const;
+
+  // inserts an item into the given slot, pushing all subsequent keys forward
+  // @note if the item is not embedded in key, shift the right half as well
+  void insert_at(unsigned slot, const ghobject_t& oid, const item_t& item);
+  // used by InnerNode for updating the keys indexing its children when their lower boundaries
+  // is updated
+  void update_key_at(unsigned slot, const ghobject_t& oid);
+
+  // @return the number of element to grab
+  std::pair<unsigned, uint16_t> calc_grab_front(uint16_t min_grab, uint16_t max_grab) const;
+  std::pair<unsigned, uint16_t> calc_grab_back(uint16_t min_grab, uint16_t max_grab) const;
+  template<int LeftN, class Mover> void grab_from_left(
+    node_t<BlockSize, LeftN, NodeType>& left,
+    unsigned n, uint16_t bytes,
+    Mover& mover);
+  // have to define a set of trap functions
+  template<int LeftN, class Mover, ntype_t O = flip_ntype(NodeType)>
```

Comment:
I think I'm missing something, why do you need these poison placeholders?  Wouldn't calling leaf.grab_from_left(..., interior) fail to find a matching signature without them?

Reply:
no, the compiler does not deduce that the invariant of the structure of a b+ tree (or its variants) that the siblings of a node should always be the same type. say, an inner node's sibling should always be a sibling. or in another words, i failed to teach the typing system to understand this. as what i do in general is (in pseudo code)
```
// assuming i am a leaf node
// try to push to my left sibling as i am overflow
left = parent.load_child(whoami - 1);
// ^ left can be a leaf or an inner node, 
//   as the type information of "i am a leaf" is not passed to load_child()
```

probably i should add a helper like `load_sibling()` to pass this type information, so `load_child<ExpectedNodeType>()` can fail earlier withe better error information.

what do you think?

Reply:
I'm not sure I follow.  These deductions happen only on staticly known types.  At some point, you'll have to assert that the sibling is a particular type -- either with your load_child helper or not.  You'll then call grab_from_leaf on it.  I don't see how the compiler couldn't know the concrete type of that argument.

Reply:
> you'll have to assert that the sibling is a particular type

yes. see https://github.com/tchaikov/ceph/commit/289834c726a88f6eddc7c6a1b6aad4dee6fb215e#diff-31b6a4983be200b0524f43d2735cbc1bR697-R728

i am quoting these helpers here as a reference
```c++
  // helper to create a node
  template<int ThisN, int MaxN>
  static child_node_t load_node_at(int n, bool is_leaf,
                                   TransactionManager* tm,
                                   OnodeBlock::Ref extent)
  {
    static_assert(ThisN < MaxN);
    if (n == ThisN) {
      if (is_leaf) {
        return LeafNode<BlockSize, ThisN>(tm, std::move(extent));
      } else {
        return InnerNode<BlockSize, ThisN>(tm, std::move(extent));
      }
    } else if constexpr (ThisN + 1 < MaxN) {
      return load_node_at<ThisN + 1, MaxN>(n, is_leaf, tm, std::move(extent));
    } else {
      // bad tag
      throw std::logic_error(fmt::format("bad node level={}, leaf={}", n, is_leaf));
    }
  }

  read_ertr::future<child_node_t>
  load_child(unsigned slot, Transaction& txn) const {
    static_assert(N < 4);
    return this->load_block(slot, txn).safe_then([this](OnodeBlock::Ref block) {
    // the L of child nodes can only be L, L.., 3
      auto p = reinterpret_cast<uint8_t*>(block->get_bptr().c_str());
      tag_t tag{*p};
      return read_ertr::make_ready_future<child_node_t>(
        load_node_at<N, 4>(tag.layout(), tag.is_leaf(), this->tm, std::move(block)));
    });
  }
```

> I don't see how the compiler couldn't know the concrete type of that argument.

right. i was not saying that it's not possible or the compiler couldn't know the type. i was just wondering if i should add another helper to enforce the node type. but per your reply, i guess the answer is "why not?".

Reply:
It's fine either way for now, it'll become clear with the next wave.